go/enclave/core/batchlimiter.go

```diff
@@ -0,0 +1,107 @@
+package core
+
+import (
+	"errors"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/rlp"
+)
+
+// BatchSizeLimiter - Acts as a limiter for batches based
+// the data from the transaction that we have to publish to the l1.
+// Acts as a calldata reservation system that accounts for both
+// transactions and cross chain messages.
+type batchSizeLimiter struct {
+	Size            uint64
+	ContractAddress common.Address
+	Topic           common.Hash
+}
+
+type BatchSizeLimiter interface {
+	AcceptTransaction(tx *types.Transaction) error
+	ProcessReceipt(receipt *types.Receipt) error
+}
+
+var ErrInsufficientSpace = errors.New("insufficient space in BatchSizeLimiter")
+
+// BatchMaxTransactionData - The number where we will cut off processing transactions inside the evm facade.
+const BatchMaxTransactionData = 25_000
+
+// NewBatchSizeLimiter - Size is the total space available per batch for calldata in a rollup.
+// contractAddress - the address of the l2 message bus where cross chain events would originate from.
+// topic - the event id of the cross chain message event.
+func NewBatchSizeLimiter(size uint64, contractAddress common.Address, topic common.Hash) BatchSizeLimiter {
+	return &batchSizeLimiter{
+		Size:            size,
+		ContractAddress: contractAddress,
+		Topic:           topic,
+	}
+}
+
+// AcceptTransaction - transaction is rlp encoded as it normally would be when publishing a rollup and
+// its size is deducted from the remaining limit.
+func (l *batchSizeLimiter) AcceptTransaction(tx *types.Transaction) error {
+	rlpSize, err := getRlpSize(tx)
+	if err != nil {
+		return err
+	}
+
+	if uint64(rlpSize) > l.Size {
+		return ErrInsufficientSpace
+	}
+
+	l.Size -= uint64(rlpSize)
```

Comment:
Should Size be a big ? And should we use the big safe ops that avoid over and underflows, or maybe it's overkill ?

Reply:
This is calldata size, definitely overkill